### ISSUE TITLE
Enable reflog view to be searchable and filterable by displaying it in main window

### ISSUE THREAD

**Issue by johndog** [gitextensions/gitextensions#5939]
Feature Request.

Currently, there is a 'show reflog' command in Main Menu -> Commands -> Show reflog… 

This command brings up a new dialog window with a view of the reflog.  It has the following problems:

- It is modal, so UI in the main window can't be used until the reflog window is closed.
- It doesn't provide the ability to search
- Commits can't be examined

My suggestion is to provide a mode for the main window to show the reflog there, so that:

- commits can be browsed, 
- they can be annotated with branch names when appropriate
- the rest of the GitEx UI isn't blocked by modality of the dialog

Since multiple reflogs exist, it should be possible to allow right-clicking on a branch and offering a "show reflog" command.  In that mode, the branch in the side pane can bolden the branch and show e.g.  "**master** [reflog]" to illustrate reflog mode.

**Comment by vbjay** [comment]
Umm. Look at the view menu.

**Comment by gerhardol** [comment]
![image](https://user-images.githubusercontent.com/6248932/50121892-8e58b080-025a-11e9-97e8-2551bc97726c.png)

Reflog in general is a hidden feature, especially since workflows with rewriting histories are encouraged.
Maybe this should be more promoted, I believe the option is in the right place already.

**Comment by RussKie** [comment]
Perhaps it is worth considering retiring the dialog altogether given a
significantly richer functionality is available in the main graph.

**Comment by johndog** [comment]
Wait, you're saying that "Show reflog references" shows all the tips that "git reflog" would show?  I thought that just added annotation like "show remotes" does.  To wit:  disabling "show remotes" still shows the remote tips, it just stops annotating them.  But "show reflog references" will show commits only reachable from the reflog?

**Comment by johndog** [comment]
How about annotating the reflog commits with HEAD@{1...n}, as the output of the command would do?

Can you filter the results to "git reflog master"?

So many questions....

**Comment by johndog** [comment]
How about listing the reflog one-line summary instead of the first line of the commit message?  Before I opened this request I had actually enabled "Show reflog references", but I didn't see a change... I was looking for "reflog-ish" entries to appear.  

I was especially looking for the "checkout (amend)" and "rebase -i"-type headings:

![image](https://user-images.githubusercontent.com/22213780/50124421-6bf27300-0219-11e9-98a3-fb7857f5e11b.png)

I have to say integrating the reflog with the live graph is pretty awesome -- it's more than I was intending to suggest (I was thinking in terms of a reflog mode for the main window).  But there are some useful reflog features that aren't currently surfaced.

**Comment by vbjay** [comment]
![image](https://user-images.githubusercontent.com/3629489/50130612-62363300-024c-11e9-9830-f6ab942d3166.png)

No branch, no tag.  The SHA is 5a401a39b1acb5ea8a087a61e0b611b38ef8dfdb.  Notice it is a dangling commit.  
![image](https://user-images.githubusercontent.com/3629489/50130822-80e8f980-024d-11e9-8b91-eab45a31ce05.png)

The dialog lets you see the actual reflog with the details like this was a rebase or an amend or whatnot.  They provide different things.  
![image](https://user-images.githubusercontent.com/3629489/50131007-361bb180-024e-11e9-9f82-42305d80b216.png)

So no.  We should not remove the dialog.  The dialog does the job of telling us the actions done to the repo.  To see commits that are in the revision grid that are in the reflog, just turn on that option.  If you want to examine the reflog to find a specific thing that was done, use the dialog.

**Comment by vbjay** [comment]
With reflog view turned on
![image](https://user-images.githubusercontent.com/3629489/50131484-4b91db00-0250-11e9-9593-12073e064e87.png)

Turned off
![image](https://user-images.githubusercontent.com/3629489/50131497-5cdae780-0250-11e9-9857-ddcef6f63983.png)

**Comment by vbjay** [comment]
If you just want dangling commits, use 
![image](https://user-images.githubusercontent.com/3629489/50131566-c955e680-0250-11e9-8f81-23218a90757b.png)

**Comment by vbjay** [comment]
```bash
git fsck-objects --no-progress --unreachable --no-reflogs --name-objects 2>/dev/null | cut -f3 -d' '|xargs -n1 -I {} git log --oneline -n1 {}| grep line
```

Alter the log command line to output what you want.  Alter the grep to find what you want.  Add a | to send the output to whatever command you want.

**Comment by johndog** [comment]
@vbjay: I see that the reflog commits are there in the graph when "show reflog" is enabled... I didn't recognize them for what they were before.  

The fsck-objects command pipelines you suggest are a marvel, thank you for sharing them.  Are you suggesting them for when the UI doesn't do what I need, or as a means of implementing it in the UI?  I think I'd use those commands if I were desperate, but for the most part, running "git reflog" and using the console pager to search comments does well enough.  Integrating that capability into the GUI would make it a lot more powerful and efficient and enable me to do it more often.

As such, I think the main window "show reflog refs" feature would be more useful if it showed all the commit properties you get in "git reflog".  Until that happens, the "show reflog…" does offer something different.  However, I am not sure I will ever find that difference useful because it can't be searched in any way, so I for one won't lobby to keep it.

**Comment by vbjay** [comment]
I disagree. Our main window is for showing commits. Reflog is a different
animal. Use the reflog window to see reflog data.

**Comment by johndog** [comment]
Here's another idea (albeit probably a lot harder):  provide a dockable version of the reflog window, and sync the selection with the main window.

**Comment by vbjay** [comment]
1.  Is what I disagree with.
2. I can see double clicking row or sha's being links to open the commit
dialog just like double clicking a commit in the main window does now.
3. Never give up.  :D.  You can copy sha and in main window navigate to
commit as of now, of course the view option must be turned on to see a
dangling commit.

**Comment by vbjay** [comment]
Because we run git log with different arguments.  We do not always have reflog info.  The reflog is x happened then y happened then ....  The commits show something different.  If you want reflog details, use the dialog.  If you want commit info, use the main window.  The main window does let you use --walk-reflogs option to see the commits that are referenced by the reflog(including dangling).   

Non modal and raising some command to navigate to selected commit would be a great idea( maybe even turn on view reflog option automatically if this feature is used.

**Comment by johndog** [comment]
> Because we run git log with different arguments.  We do not always have reflog info.

Okay, I can see that, but this strikes me as an outer-join problem, not an architectural deal-breaker.

> If you want reflog details, use the dialog. If you want commit info, use the main window.

Here's the scenario that motivated me finally opening this request:

1. Search the reflog for commits containing the word "refactor"
2. Examine the commit.  
3. If you've found what you're looking for, you're done.
4. Go to next commit matching search in 1, goto 2

Using "git reflog" in the console for step 1, I was actually frustrated because a "/refactor" search in the pager (less?) turned out to be case sensitive, so I missed the commits with "Refactor".  Conclusion:  this would really benefit from a user-friendly GUI solution.

The scenario appears to align well with the capabilities of the main window, because it is word-for-word already supported if you replace "reflog" with "git log".  I can imagine reasons that it would be better to replicate the required functionality in the reflog window, but the question that should follow is "what will have to be replicated next?"  Today I'm filtering based on message, tomorrow I might filter on content.  Does it make sense to duplicate the whole filter/search framework?  It's certainly plausible, I don't know enough about the code to have an opinion.  

The duplication of functionality might stop with filtering and search.  If the dialog is non-modal and you can pivot back and forth between viewing commits and stepping through reflog, then being able to locate commits in the reflog is the only class of functionality the reflog window would need to support -- everything else could be provided by the main window (presumably with no changes).  

So if the reflog window is non-modal, enables search/filter, and coordinates with the main window, I think that satisfies any inclination I'd have to integrate reflog with the main window directly.  So, what do you think of the need to add search/filter to the reflog window... does that affect your opinion?

**Comment by johndog** [comment]
I have to say, "show reflog references" is awesome.  I never have to proactively create temp branches anymore... I can hard-reset at any time and decide retroactively to do something with abandoned dangling commits.  I used to create temp branches every time I rebased so that I could evaluate the integrity of my merges with a before/after comparison.  No longer necessary.

**Comment by gerhardol** [comment]
Any specific aspects of reflogs to add or can the issue be closed?

**Comment by vbjay** [comment]
I think the take away from this is it has room for improvement.

- Make dialog non modal
- Add functionality so user can select a reflog entry and navigate to that commit in the grid.  Doing so turns on option to show reflog entries so they don't try to navigate to a non existing row in the grid
